### ISSUE TITLE
Illegal character in authority

### ISSUE THREAD

**Issue by wicope** [git-commit-id/git-commit-id-maven-plugin#519]
```
			<plugin>
				<groupId>pl.project13.maven</groupId>
				<artifactId>git-commit-id-plugin</artifactId>
				<version>4.2.0</version>
			</plugin>
```

```
git clone --single-branch --branch "${BRANCH}" https://"${USERNAME}":"${PASSWORD}"@xxxx/xxxx.git
```

```
08:22:51  [ERROR] Something went wrong to strip the credentials from git's remote url (please report this)!
08:22:51  java.net.URISyntaxException: Illegal character in authority at index 8: https://xxxxxx_xxx:xE':k]{7FtB!xxxx@xxxx/xxxx.git
```

**Comment by TheSnoozer** [comment]
Hello,
thank you for reporting. I can reproduce the issue. The root issue (or at least how I can reproduce it) seems that there are special characters in the URL. The special characters might or might not be used in the password that we want to filter out. 
As per https://www.ietf.org/rfc/rfc2396.txt only a certain subset of characters might be allowed in certain parts of the URL. 

A few examples:
```
            // use of 'reserved' characters as https://www.ietf.org/rfc/rfc2396.txt
            { "https://user:A;/?:@&=+$,Z@example.com:8888", "https://user@example.com:8888" },
            // use of 'unreserved' characters as https://www.ietf.org/rfc/rfc2396.txt
            { "https://user:A-_.!~*'()Z@example.com:8888", "https://user@example.com:8888" },
            // use of 'delims' characters as https://www.ietf.org/rfc/rfc2396.txt
            { "https://user:A<>#%\"Z@example.com:8888", "https://user@example.com:8888" },
            // use of 'unwise' characters as https://www.ietf.org/rfc/rfc2396.txt
            { "https://user:A{}|\\^[]`Z@example.com:8888", "https://user@example.com:8888" },
```
For almost all the instances mentioned here I'm pretty confident that I have a working fix already (essentially encoding the special characters). However for the `/` character (part of the reserved characters) the situation grows more complex. In general when given a `https://"${USERNAME}":"${PASSWORD}"` URL there is no guarantee that the password doesn't contain a `/`. Now blindly replacing the `/`, would mess with file-based like URLs (e.g. `/path/to/repo.git/`). Doing extra work of finding the ones that need replacement sounds somewhat overkill as well (note that password could have `@` and `:` symbols so any assumption where the hostname starts might be just a good guess). So a current conclusion I'd say most of the characters should be handled, only the `/` would yield in a similar issue.

Note that in the end this is not a security problem. The plugin doesn't expose any password. If the extraction fails for whatever reason the entire URL will be reported as empty string.

**Comment by TheSnoozer** [comment]
Thanks again for reporting! The plugin now tries to a bit smarter in terms of escaping, however we are still bound what the java URI implementation accepts. In a nutshell this is defined by https://www.ietf.org/rfc/rfc2396.txt that declares a valid URL schema as:
```
<userinfo>@<host>:<port>

userinfo = *( unreserved | escaped | ";" | ":" | "&" | "=" | "+" | "$" | "," )

unreserved  = alphanum | mark
mark = "-" | "_" | "." | "!" | "~" | "*" | "'" | "(" | ")"
alphanum = alpha | digit
digit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" |  "8" | "9"
alpha = lowalpha | upalpha
lowalpha = "a" | "b" | "c" | ... | "x" | "y" | "z"
upalpha = "A" | "B" | "C" | ... | "X" | "Y" | "Z"
```

In a nutshell this means everything not covered here, needs escaping (e.g. the `@`-sign https://stackoverflow.com/a/29356143/7890667). This can also simply be verified by running the following:
```
$ git clone http://user:passw@rd@localhost
Cloning into 'localhost'...
fatal: unable to access 'http://user:passw@rd@localhost/': Could not resolve host: rd@localhost
$ git clone http://user:passw%40rd@localhost
Cloning into 'localhost'...
fatal: unable to access 'http://user:passw%40rd@localhost/': Failed to connect to localhost port 80: Connection refused
```
Now the plugin can not blindly replace all `@` with the proper escaping and would yield the same issue as outlined in here. However the URL itself doesn't follow the https://www.ietf.org/rfc/rfc2396.txt-standard which makes it feel ok when the replacement fails. In case not mentioned: If anything 'fails' the remote URL is not exposed as property simply meant as save-guard to avoid that a password might slip into a property file. 

Overall I agree with the following sentiment outlined in the RFC:
```
   Some URL schemes use the format "user:password" in the userinfo
   field. This practice is NOT RECOMMENDED, because the passing of
   authentication information in clear text (such as URI) has proven to
   be a security risk in almost every case where it has been used.
```
I would strongly advice to use alternatives (ssh-keys).

TLDR:
- The plugin was made slightly smarter by escaping some special characters in URLs
- However the plugin can't blindly escape them all (e.g. the `@` character) which when used in it#s raw form as the password inside the URL would yield the same problem as outlined here
- By definition of [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt) a raw `@` character is not allowed inside the `userinfo`/`authority` section and would needed to be escaped with `%40`, hence the plugin would only choke on non [RFC 2396](https://www.ietf.org/rfc/rfc2396.txt) complaint URLs. 


For me that sounds good enough as a fix. Thanks again for reporting! Closing now as 'partially fixed'
